### PR TITLE
Eye targeting is not working with near hand pointers

### PR DESCRIPTION
**Pull request by wiwei**
https://github.com/microsoft/MixedRealityToolkit-Unity/issues/4431

In a previous change (https://github.com/microsoft/MixedRealityToolkit-Unity/pull/4270) a gaze provider state machine was added to bring the head gaze pointer behavior in-line with how the shell works. This had the side effect of also impacting eye-gaze state. This change brings the old eye-gaze behavior back by updating the state machine to be aware of the type of gaze being provided.

Also added a bunch of tests for this new thing, along with tests that show that transitions between the two (for example, when the eye tracking system gains tracking or loses tracking) works well.

### CODE REVIEW

File: Assets/MixedRealityToolkit.Services/InputSystem/GazePointerVisibilityStateMachine.cs [hunk 1]
```diff
@@ -4,31 +4,76 @@
...
+
+            gazePointerState = isEyeGazePointerActive ?
+                GazePointerState.GazePointerActive :
+                GazePointerState.GazePointerInactive;
+        }
+
+        private void UpdateStateHeadGaze(int numNearPointersActive, int numFarPointersActive)
         {
-            GazePointerState newState = gazePointerState;
             bool isMotionControllerOrHandUp = numFarPointersActive > 0 || numNearPointersActive > 0;
```
Comment (julenka) [inline Assets/MixedRealityToolkit.Services/InputSystem/GazePointerVisibilityStateMachine.cs]: This reminds me of bug #4405. Mind pulling that change in? It's very short, just replace with 

> bool canGazeCursorShow = numFarPointersActive == 1 && numNearPointersActive == 0;

File: Assets/MixedRealityToolkit.Tests/EditModeTests/InputSystem/GazePointerStateMachineTests.cs [hunk 2]
```diff
@@ -34,16 +34,122 @@ public void TestHandAndSpeechBehaviour()
...
+            TestUtilities.InitializeMixedRealityToolkitScene(true);
+
+            // Initial state: gaze pointer active
+            var gsm = new GazePointerVisibilityStateMachine();
+            Assert.IsTrue(gsm.IsGazePointerActive, "Gaze pointer should be visible on start");
+
+            // With the hand raised, eye gaze pointer should still exist because only far interaction causes the 
+            // eye gaze pointer to go away.
+            gsm.UpdateState(1 /*numNearPointersActive*/, 0 /*numFarPointersActive*/, true);
+            Assert.IsTrue(gsm.IsGazePointerActive, "With near interaction, gaze pointer should continue to exist");
```
Comment (julenka) [inline Assets/MixedRealityToolkit.Tests/EditModeTests/InputSystem/GazePointerStateMachineTests.cs]: eye gaze?

File: Assets/MixedRealityToolkit.Tests/EditModeTests/InputSystem/GazePointerStateMachineTests.cs [hunk 3]
```diff
@@ -34,16 +34,122 @@ public void TestHandAndSpeechBehaviour()
...
+            gsm.UpdateState(0 /*numNearPointersActive*/, 1 /*numFarPointersActive*/, true);
+            Assert.IsFalse(gsm.IsGazePointerActive, "With far interaction, gaze pointer should go away");
+
+            // Reset the state and validate that it goes back to being visible.
+            gsm.UpdateState(0 /*numNearPointersActive*/, 0 /*numFarPointersActive*/, true);
+            Assert.IsTrue(gsm.IsGazePointerActive, "Gaze pointer should be visible when no near or far pointers");
+
+            // Saying "select" should have no impact on the state of eye gaze-based interactions.
+            SpeechEventData data = new SpeechEventData(EventSystem.current);
+            data.Initialize(new BaseGenericInputSource("test input source", new IMixedRealityPointer[0], InputSourceType.Voice),
```
Comment (julenka) [inline Assets/MixedRealityToolkit.Tests/EditModeTests/InputSystem/GazePointerStateMachineTests.cs]: Maybe pull into utility? We've pasted this a few times now.

File: Assets/MixedRealityToolkit.Services/InputSystem/FocusProvider.cs [hunk 4]
```diff
@@ -878,7 +878,10 @@ private void ReconcilePointers()
             }
             if (gazePointer != null)
             {
-                gazePointerStateMachine.UpdateState(NumNearPointersActive, NumFarPointersActive);
+                gazePointerStateMachine.UpdateState(
+                    NumNearPointersActive,
+                    NumFarPointersActive,
+                    InputSystem.EyeGazeProvider.IsEyeGazeValid);
```
Comment (julenka) [inline Assets/MixedRealityToolkit.Services/InputSystem/FocusProvider.cs]: Can this be null ever?

File: Assets/MixedRealityToolkit.Tests/EditModeTests/InputSystem/GazePointerStateMachineTests.cs (hunk 3 above)
Comment (wiwei) [reply Assets/MixedRealityToolkit.Tests/EditModeTests/InputSystem/GazePointerStateMachineTests.cs]: Put into a static private, utility is overkill until we need to use it in multiple places.

File: Assets/MixedRealityToolkit.Services/InputSystem/FocusProvider.cs (hunk 4 above)
Comment (wiwei) [reply Assets/MixedRealityToolkit.Services/InputSystem/FocusProvider.cs]: InputSystem is a prereq to this file (i.e. other places have the same level of dependency)

EyeGazeProvider is set up by the input system uniquely within its startup path and startup will fail if there is no gaze provider given.

So tl;dr Not as far as I can see.

File: Assets/MixedRealityToolkit.Services/InputSystem/GazePointerVisibilityStateMachine.cs (hunk 1 above)
Comment (wiwei) [reply Assets/MixedRealityToolkit.Services/InputSystem/GazePointerVisibilityStateMachine.cs]: Done - I had to change up the suggested implementation a bit (i.e. #4405 had a bug in its proposed implementation in the missing negation of canGazeCursorShow within its if checks)

**Pull request merged**